# **Tarea 1 - CC6205 Natural Language Processing 📚**

**Integrantes:** Louise Schmidt y Sebastián Urbina

**Fecha límite de entrega 📆:** Lunes 05 de Abril.

**Tiempo estimado de dedicación:** 6 horas.




` ` 


Bienvenid@s a la primera tarea en el curso de Natural Language Processing (NLP). Esta tarea tiene como objetivo evaluar los contenidos teóricos de las primeras semanas de clases, enfocado principalmente en ***Information Retrieval (IR)*** y ***Vector Space Models***. Si aún no han visto las clases, se recomienda visitar los links de las referencias.

La tarea consta de una parte teórica que busca evaluar conceptos vistos en clases. Seguido por una parte práctica con el fín de introducirlos a la programación en Python enfocada en NLP. 

` ` 




**Instrucciones:**
- La tarea se realiza en grupos de **máximo** 2 personas. Puede ser invidivual pero no es recomendable.
- La entrega es a través de u-cursos a más tardar el día estipulado arriba. No se aceptan atrasos.
- El formato de entrega es este mismo Jupyter Notebook.
- Al momento de la revisión tu código será ejecutado. Por favor verifica que tu entrega no tenga errores de compilación. 
- Está **PROHIBIDO** usar cualquier librería que implemente los algoritmos pedidos (Spacy, scikit, etc). Sólo se podrán utilizar las librerías importadas al inicio de la sección de práctica.
- En el horario de auxiliar pueden realizar consultas acerca de la tarea a través del canal de Discord del curso. 



Ahora sí, empecemos! 😄🚀

` ` 

**Referencias:**


Slides:
    
- [Introducción al curso](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-introduction.pdf)
- [Vector Space Model / Information Retrieval](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-IR.pdf)    

Videos: 

- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte I](https://www.youtube.com/watch?v=HEKTNOttGvU)
- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte II](https://www.youtube.com/watch?v=P8cwnI-f-Kg)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 1](https://youtu.be/FXIVClF370w)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 2](https://youtu.be/f8nG1EMmPZk)

` ` 



## **1 - Preguntas teóricas 📕 (2 puntos).** ##









Las siguientes celdas contienen preguntas acerca del contenido visto en clases y en el material del curso.  Contestar cada pregunta en su celda correspondiente y **no extenderse más de 5 lineas** . 🙏
` ` 

**Pregunta 1 (0.5 puntos): ¿Por qué el análisis del lenguaje humano es una tarea compleja? Mencione dos razones.**



` ` 

Dos razones por las cuales el leguanje humano es una tarea compleja son:

1- Es altamente ambiguo. Por ejemplo, podemos tener frases con una leve diferencia de una o dos palabras y el propósito de la frase cambia totalmente.

2- Cambia y evoluciona constantemente. Por ejemplo, el idioma chileno que hablamos actualmente evolucióno desde el español y el latín.  

` ` 

**Pregunta 2 (0.5 puntos): ¿Cuál es la diferencia entre Natural Language Processing (NLP) y Computational Linguistics?**

` ` 

NLP se enfoca en encontrar métodos para resolver problemas prácticos que involucran el leguaje humano. Se construye un sistema que resuelve un problema específico. Por ejemplo, extraer datos de un texto, traducir textos. 

Por otro lado, Computational Linguistics está más asociado con la linguística y comprender el lenguaje humano ayudándose a través de la computación. Por ejemplo, cómo se usa el lenguaje, cómo se aprende, si dos lenguas estan relacionadas entre sí. No busca resolver un problema específico.

` ` 

**Pregunta 3 (0.5 puntos): ¿Qué es el Machine Learning supersivado? ¿Por qué es importante para el campo de NLP?**

` ` 

El Machine Learning supervisado es utilizar un algoritmo que aprende a generar un output a partir de un input. Es decir, tengo una "respuesta correcta" que la máquina produce automáticamente. El input y el output correctos están dados en un "training dataset". 
Es muy útil debido a que la máquina encuentra las reglas con los inputs y outputs. 

` ` 

**Pregunta 4 (0.5 puntos): ¿Cuáles son las diferencias entre usar Deep learning y Machine Learning clásico (empirismo) para un problema de NLP? Ejemplifique con alguna task.**

` ` 

En el caso de Machine Learning habrá una persona que genera una clasificación de cierto vocaublario o símbolos, lo cual puede tomar bastante tiempo. Luego, cuando se crea esto, se entregan los datos preprocesados con características extraidas manualmente son clasificados por la máquina generando un output. Por otro lado, en Deep Learning la máquina extrae los atributos del input de manera automática, permitiendo reconocer patrones de forma más eficiente y clasificando el input, generando el output. A grandes rasgos el Deep Learning permite identificar patrones lingüisticos de forma automática, evitando la realización de características de forma manual para alimentar un algoritmo, como se realiza en Machile Learning clásico.

Por ejemplo: 

- Sistemas de preguntas y resuesta o ChatBot: Si se utilizara Machine Learning Clásico sería bastante tedioso intentar cubrir todas las opciones posibles, por el simple hecho que habría que clasificar una gran gama de posibilidades y puede que el modelo no sea generalizable, sin embargo con Deep Learning se pueden implementar este tipo de tareas de manera más óptima con redes neuronales pre-entrenadas, las cuales se comportan bastante bien y se usan actualmente.
` ` 

## **2 - Preguntas prácticas 💻 (4 puntos).** ##

Esta segunda sección incluye ejercicios de programación 🤙. Leer atentamente las instrucciones entregadas a continuación para facilitar el proceso de revisión de sus trabajos.


` ` 

**Instrucciones:**



- Escribe tu código entre las lineas de comentarios **### Aquí inicia tu código ###** y **### Aquí termina tu código ###**.
- Cuando el ejercicio incluya un bloque llamado ***Test***, comprueba que el resultado de la ejecución coincida con el resultado esperado.
- Recuerde siempre mantener buenas prácticas de código.
- Está permitido sólo utilizar las librerías importadas antes del Ejercicio 1.
- **Recordar** que: *Documento = Oración. Dataset = Corpus. Vocabulario = Typos*.




**Ejemplo:** Implemente una función **`hello_world()`** que imprima en pantalla `"Hello World"`. 

In [76]:
def hello_world():
  ### Aquí inicia tu código ###
  print("Hello World")
  ### Aquí termina tu código ###

***Test:***

In [77]:
hello_world()

Hello World


***Resultado esperado***: 
<table>
    <tr> 
        <td> Hello World </td> 
    </tr>
</table> 

``
``




Estas son las librerías permitidas. Si quieren utilizar alguna librería adicional, pueden realizar la consulta a través de Discord. 

In [78]:
import codecs
import re
import numpy as np
import pandas as pd

` ` 

**Ejercicio 1 - *Tokenización* (0.5 puntos).** 
` `  
` ` 

En el primer ejercicio veremos la dificultad 😨 de tokenizar textos no estructurados, destacando la importancia de tener librerías que realicen este trabajo. 

El archivo adjunto al enunciado de la tarea contiene una interconsulta médica real en Chile, estos textos son frecuentemente utilizados para obtener información epidemiológica y realizar estudios estadísticos a partir de ellos. Es por esto que poder analizar el texto de manera correcta es fundamental. 


Ejecute el código a continuación para cargar el ejemplo. Recuerde realizar la modificación al directorio en caso que el archivo no se encuentre en el mismo directorio del Jupyter Notebook.

In [79]:
text = codecs.open('interconsulta.txt', 'r', 'UTF-8').read()
print(text)

Interconsulta n°2 del mes.-05/10/2019
- ENFERMEDAD RENAL  CRÓNICA ETAPA 3. 
- Fundamento Clínico APS: PACIENTE  DE 61 AÑOS, 	CON ANTECEDENTES DE HRTA DE LARGA DATA.ACUDE A CONTROL PSCV CON EXAMNES, DONDE DESTACA CREATININA DE 1,43,
 CON UNA VFG CALCULADA DE 53 ML/MIN/1.73M2. CON UNA CREATININA HACE MAS DE 3 MESES DE 1.34.
                   
Enfermedad renal cronica Etapa 3(SOLICITO EVALUACION DEL CASO).



Implementen una función **`get_tokens()`** que reciba un texto y entregue una lista con sus tokens. Son libres de elegir la forma de tokenizar mientras no utilicen librerías con tokenizadores ya implementados. Pueden utilizar la librería **re** importada para trabajar símbolos.


Ejemplo de uso:

`get_tokens('Este es un ejemplo de prueba.')` 

Nos entregaría:

`['Este', 'es', 'un', 'ejemplo', 'de', 'prueba', '.']`

In [80]:
import re #Utilizamos la librería RegEx(re)

def get_tokens(text):
  ### Inicio del código ###
  tokens = re.findall("[\w]+|[\S]", text) #Buscamos todos los string que sean palabras ó caracteres que no sean espacios
  return tokens
  ### Fín del código ###
get_tokens('Este es un ejemplo de prueba.')

['Este', 'es', 'un', 'ejemplo', 'de', 'prueba', '.']

In [81]:
tokens = get_tokens(text)
tokens

['Interconsulta',
 'n',
 '°',
 '2',
 'del',
 'mes',
 '.',
 '-',
 '05',
 '/',
 '10',
 '/',
 '2019',
 '-',
 'ENFERMEDAD',
 'RENAL',
 'CRÓNICA',
 'ETAPA',
 '3',
 '.',
 '-',
 'Fundamento',
 'Clínico',
 'APS',
 ':',
 'PACIENTE',
 'DE',
 '61',
 'AÑOS',
 ',',
 'CON',
 'ANTECEDENTES',
 'DE',
 'HRTA',
 'DE',
 'LARGA',
 'DATA',
 '.',
 'ACUDE',
 'A',
 'CONTROL',
 'PSCV',
 'CON',
 'EXAMNES',
 ',',
 'DONDE',
 'DESTACA',
 'CREATININA',
 'DE',
 '1',
 ',',
 '43',
 ',',
 'CON',
 'UNA',
 'VFG',
 'CALCULADA',
 'DE',
 '53',
 'ML',
 '/',
 'MIN',
 '/',
 '1',
 '.',
 '73M2',
 '.',
 'CON',
 'UNA',
 'CREATININA',
 'HACE',
 'MAS',
 'DE',
 '3',
 'MESES',
 'DE',
 '1',
 '.',
 '34',
 '.',
 'Enfermedad',
 'renal',
 'cronica',
 'Etapa',
 '3',
 '(',
 'SOLICITO',
 'EVALUACION',
 'DEL',
 'CASO',
 ')',
 '.']

**Describa cuáles fueron sus supuestos para realizar la tokenización y compare sus tokens con los entregados por la librería nltk en el bloque de código de más abajo.**

Se utilizó la librería "RegEx" de python para identificar palabras y caracteres especiales que no sean espacios("\n","\t","\r"). Para esto se definió el criterio de búsqueda en base a la sintaxis de "Re" como `"[\w]+|[\S]"`. Lo cual busca palabras completas([\w]) ó(|) caracteres que no sean espacios([\S]).

La única diferencia que se encontró respecto a la librería `nltk` fueron "-." y ").". Los cuales a diferencia del tokenizador definido en base a `Re` los identificó como caracteres independientes, es decir, "-", ".", ")", ".".

In [82]:
from nltk.tokenize import wordpunct_tokenize 
nltk_tokens = wordpunct_tokenize(text)
nltk_tokens

['Interconsulta',
 'n',
 '°',
 '2',
 'del',
 'mes',
 '.-',
 '05',
 '/',
 '10',
 '/',
 '2019',
 '-',
 'ENFERMEDAD',
 'RENAL',
 'CRÓNICA',
 'ETAPA',
 '3',
 '.',
 '-',
 'Fundamento',
 'Clínico',
 'APS',
 ':',
 'PACIENTE',
 'DE',
 '61',
 'AÑOS',
 ',',
 'CON',
 'ANTECEDENTES',
 'DE',
 'HRTA',
 'DE',
 'LARGA',
 'DATA',
 '.',
 'ACUDE',
 'A',
 'CONTROL',
 'PSCV',
 'CON',
 'EXAMNES',
 ',',
 'DONDE',
 'DESTACA',
 'CREATININA',
 'DE',
 '1',
 ',',
 '43',
 ',',
 'CON',
 'UNA',
 'VFG',
 'CALCULADA',
 'DE',
 '53',
 'ML',
 '/',
 'MIN',
 '/',
 '1',
 '.',
 '73M2',
 '.',
 'CON',
 'UNA',
 'CREATININA',
 'HACE',
 'MAS',
 'DE',
 '3',
 'MESES',
 'DE',
 '1',
 '.',
 '34',
 '.',
 'Enfermedad',
 'renal',
 'cronica',
 'Etapa',
 '3',
 '(',
 'SOLICITO',
 'EVALUACION',
 'DEL',
 'CASO',
 ').']

` `  
` ` 

**Ejercicio 2 - *Stopwords y Stemming* (1 punto).** 
` `  
` ` 

Considere el siguiente corpus:


In [83]:
dataset = ["I like human languages", "I like programming languages", "Spanish is my favorite language"]

Diseñe una función **`get_vocab()`** que extraiga los typos de este corpus solamente tokenizando. Puede utilizar la función del Ejercicio 1.

In [84]:
def get_vocab(dataset):
  ### Aquí inicia tu código ###
  vocab = []
  for sentence in dataset: #Recorremos las oraciones del corpus
      tokens = get_tokens(sentence) #Tokenizamos la oración
      for token in tokens: #Recorremos cada token
          if token not in vocab: #Nos aseguramos de almacenar los tokens únicos
              vocab.append(token)
  return vocab
  ### Aquí termina tu código ###

***Test:***

In [85]:
vocab = get_vocab(dataset)
vocab

['I',
 'like',
 'human',
 'languages',
 'programming',
 'Spanish',
 'is',
 'my',
 'favorite',
 'language']

``
``

***Resultado esperado***: 
<table>
    <tr> 
        <td>['favorite',
 'Spanish',
 'language',
 'I',
 'like',
 'programming',
 'languages',
 'my',
 'human',
 'is'] </td> 
    </tr>
</table> 

``
``

Ahora diseñe reglas que usted estime convenientes tanto de **Stemming** como de **Stopwords**. Implemente una función que reciba una lista con los elementos del vocabulario, le aplique sus reglas y devuelva el vocabulario actualizado. 

    Explique sus reglas aquí

    - Definir lista de stopWords que no aportan mucho información, para este caso "i", "is", "my". 
    - Filtrar caracteres especiales, dejando sólo palabras
    - Recortar palabras terminadas en "s", asumiendo que en general se asocian a términos en plural.


In [87]:
stopwords = ["i","is","my"]

def pre_processing(vocab):
    ### Aquí inicia tu código ###
    for i, word in enumerate(vocab):
        if not word.isalpha():
            vocab.remove(word)
        if word.lower() in stopwords:
            vocab.remove(word)
        if word[-1] == "s":
            vocab[i] = word[0:-1]
    vocab_final = list(np.unique(vocab))
    return vocab_final
    ### Aquí termina tu código ###

In [88]:
vocab = pre_processing(vocab)
vocab

['Spanish', 'favorite', 'human', 'i', 'language', 'like', 'programming']

` `  
` ` 

**Ejercicio 3 - *Bag of Words* 🐶🐈(0.5 puntos).** 
` `  
` ` 

Considere el siguiente corpus, donde cada elemento del arreglo representa un documento:

In [89]:
d0 = 'El perro se come la comida y después se duerme'
d1 = 'El perro se despierta y después empieza a ladrar'
d2 = 'El perro ladra y después se come la comida'
d3 = 'El gato se come la comida y después se duerme'
d4 = 'El gato se despierta y después empieza a maullar'
d5 = 'El gato maulla y después se come la comida'
dataset = [d0, d1, d2, d3, d4, d5]

El objetivo de este ejercicio es determinar cuáles de  los documentos entregados son los más similares entre sí. Para ello utilizaremos la técnica TF-IDF. 

Como los algoritmos de Machine Learning no comprenden el texto en lenguaje natural, estos documentos deben ser convertidos a vectores numéricos. La representación más simple vista en clases es el **Bag of Words**, método mediante el cuál se cuentan las apariciones de cada palabra en cada uno de los documentos entregados.

Implemente la función **`bag_of_words()`**, que reciba como input un arreglo de documentos y devuelva un pandas dataframe con la representación Bag of Words de los documentos entregados. En esta representación las columnas son el vocabulario y las filas representa las apariciones de cada una de las palabras en los documentos. En otras palabras, cada fila representa el bow de un determinado documento.


Por ejemplo para el siguiente dataset: 

```
dataset = ['El perro ladra', 'El perro come']
```

Debiese entregarnos lo siguiente:


|   | el | perro | ladra | come |
|---|----|-------|------|-------|
| 0 | 1  | 1     | 1    | 0     |
| 1 | 1  | 1     | 0    | 1     |



In [90]:
def bag_of_words(dataset):
    ### Aquí inicia tu código ###
    vocab = get_vocab(dataset) #Obtenemos los types
    bag = []
    for sentence in dataset:
        tokens = get_tokens(sentence) 
        vec = np.zeros(len(vocab), dtype = int) #Creamos un vector del largo de los types para almacenar las recuencias
        for i, word in enumerate(vocab): 
            word_freq = 0
            for token in tokens:
                if word == token:
                    word_freq += 1
            vec[i] = word_freq

        bag.append(vec)
    df = pd.DataFrame(bag, columns = vocab, index = ['d'+str(i) for i in range(len(dataset))])
    return df
    ### Aquí termina tu código ###

***Test:***

In [91]:
dataset_bow = bag_of_words(dataset)
dataset_bow

El  perro  se  come  la  comida  ...  a  ladrar  ladra  gato  maullar  maulla
d0   1      1   2     1   1       1  ...  0       0      0     0        0       0
d1   1      1   1     0   0       0  ...  1       1      0     0        0       0
d2   1      1   1     1   1       1  ...  0       0      1     0        0       0
d3   1      0   2     1   1       1  ...  0       0      0     1        0       0
d4   1      0   1     0   0       0  ...  1       0      0     1        1       0
d5   1      0   1     1   1       1  ...  0       0      0     1        0       1

[6 rows x 17 columns]

``
``

***Resultado esperado***: 

|    | El | perro | se | come | la | comida | y | después | duerme | despierta | empieza | a | ladrar | ladra | gato | maullar | maulla |
|----|---:|------:|---:|-----:|---:|-------:|--:|--------:|-------:|----------:|--------:|--:|-------:|------:|-----:|--------:|-------:|
| d0 |  1 |     1 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    0 |       0 |      0 |
| d1 |  1 |     1 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      1 |     0 |    0 |       0 |      0 |
| d2 |  1 |     1 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     1 |    0 |       0 |      0 |
| d3 |  1 |     0 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      0 |
| d4 |  1 |     0 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      0 |     0 |    1 |       1 |      0 |
| d5 |  1 |     0 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      1 |

``
``

` `  
` ` 

**Ejercicio 4 - *Calcular TF* (0.5 puntos):** Ahora debemos usar el dataframe del ejercicio anterior para calcular la matriz de TF normalizada por la máxima frecuencia ${max_i({\text{tf}_{i,j})}}$, donde
i corresponde al índice de las filas (bow) y j al de las columnas (palabras). Es decir, dividir cada bow en la cantidad de veces de la palabra que aparezca más veces en ese vector. 


$$\text{nft}_{i,j} = \frac{\text{tf}_{i,j}}{max_i({\text{tf}_{i,j})}}$$

In [92]:
def calc_tf(dataset_bow):
    ### Aquí inicia tu código ###
    n = dataset_bow.shape[0]
    tf = dataset_bow.copy()
    for d in range(n):
        tf.iloc[d,:] = tf.iloc[d,:] / np.max(tf.iloc[d,:]) #Dividimos cada vector fila por la máxima frecuencia correspondiente
    return tf
    ### Aquí termina tu código ###

***Test:***

In [93]:
tf = calc_tf(dataset_bow)
tf

El  perro   se  come   la  ...  ladrar  ladra  gato  maullar  maulla
d0  0.5    0.5  1.0   0.5  0.5  ...     0.0    0.0   0.0      0.0     0.0
d1  1.0    1.0  1.0   0.0  0.0  ...     1.0    0.0   0.0      0.0     0.0
d2  1.0    1.0  1.0   1.0  1.0  ...     0.0    1.0   0.0      0.0     0.0
d3  0.5    0.0  1.0   0.5  0.5  ...     0.0    0.0   0.5      0.0     0.0
d4  1.0    0.0  1.0   0.0  0.0  ...     0.0    0.0   1.0      1.0     0.0
d5  1.0    0.0  1.0   1.0  1.0  ...     0.0    0.0   1.0      0.0     1.0

[6 rows x 17 columns]

``
``

***Resultado esperado***: 

|    |  El | perro |  se | come |  la | comida |   y | después | duerme | despierta | empieza |   a | ladrar | ladra | gato | maullar | maulla |
|----|----:|------:|----:|-----:|----:|-------:|----:|--------:|-------:|----------:|--------:|----:|-------:|------:|-----:|--------:|-------:|
| d0 | 0.5 |   0.5 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d1 | 1.0 |   1.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    1.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d2 | 1.0 |   1.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   1.0 |  0.0 |     0.0 |    0.0 |
| d3 | 0.5 |   0.0 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.5 |     0.0 |    0.0 |
| d4 | 1.0 |   0.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    0.0 |   0.0 |  1.0 |     1.0 |    0.0 |
| d5 | 1.0 |   0.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  1.0 |     0.0 |    1.0 |

``
``

` `  
` ` 

**Ejercicio 5 - *Calcular IDF* (0.5 punto):**


Implementar `calc_idf(dataset_bow)`. Este debe retornar un diccionario en donde las llaves sean las palabras y los valores sean el calculo de cada idf por palabra.

Recordar que $idf_{t_i} = log_{10}\frac{N}{n_i}$ con $N = $ número de documentos y $n_i = $ Número de documentos que contienen la palabra $t_i$ 

In [94]:
def calc_idf(dataset_bow):
    ### Aquí inicia tu código ###
    N = dataset_bow.shape[0]
    vocab = dataset_bow.columns
    idf = {}
    for word in vocab:
        n_i = 0
        for d in range(N):
            if dataset_bow[word][d] != 0:
                n_i += 1
        idf[word] = np.log10(N/n_i)

    return idf    
    ### Aquí termina tu código ###

***Test:***

In [95]:
idf = calc_idf(dataset_bow)
idf

{'El': 0.0,
 'a': 0.47712125471966244,
 'come': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'despierta': 0.47712125471966244,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'y': 0.0}

``
``

***Resultado esperado***: 
<table>
    <tr> 
        <td> {'El': 0.0,
 'a': 0.47712125471966244,
 'come': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'despierta': 0.47712125471966244,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'y': 0.0} </td> 
    </tr>
</table> 

``
``

Puede notar el bajo puntaje otorgado a las palabras que más se repiten! 😮

**Ejercicio 6 - *Calcular TF-IDF & concluir similitud de documentos.* (1 punto)**


Programe la función `calc_tf_idf(tf, idf)` que entrega el dataframe TF-IDF asociado al dataset que estamos analizando.

In [96]:
def calc_tf_idf(tf, idf):
    ### Aquí inicia tu código ###
    tf_idf = tf.copy()
    for word in idf.keys():
        tf_idf[word] = tf_idf[word]*idf[word]
    
    return tf_idf
    ### Aquí termina tu código ### 

***Test.***

In [97]:
tf_idf = calc_tf_idf(tf, idf)
tf_idf

El     perro   se      come  ...     ladra      gato   maullar    maulla
d0  0.0  0.150515  0.0  0.088046  ...  0.000000  0.000000  0.000000  0.000000
d1  0.0  0.301030  0.0  0.000000  ...  0.000000  0.000000  0.000000  0.000000
d2  0.0  0.301030  0.0  0.176091  ...  0.778151  0.000000  0.000000  0.000000
d3  0.0  0.000000  0.0  0.088046  ...  0.000000  0.150515  0.000000  0.000000
d4  0.0  0.000000  0.0  0.000000  ...  0.000000  0.301030  0.778151  0.000000
d5  0.0  0.000000  0.0  0.176091  ...  0.000000  0.301030  0.000000  0.778151

[6 rows x 17 columns]

``
``

***Resultado esperado***: 

|    |  El |    perro |  se |     come |       la |   comida |   y | después |   duerme | despierta |  empieza |        a |   ladrar |    ladra |     gato |  maullar |   maulla |
|----|----:|---------:|----:|---------:|---------:|---------:|----:|--------:|---------:|----------:|---------:|---------:|---------:|---------:|---------:|---------:|---------:|
| d0 | 0.0 | 0.150515 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d1 | 0.0 | 0.301030 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.778151 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d2 | 0.0 | 0.301030 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.778151 | 0.000000 | 0.000000 | 0.000000 |
| d3 | 0.0 | 0.000000 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.150515 | 0.000000 | 0.000000 |
| d4 | 0.0 | 0.000000 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.000000 | 0.000000 | 0.301030 | 0.778151 | 0.000000 |
| d5 | 0.0 | 0.000000 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.301030 | 0.000000 | 0.778151 |


``
``

Ahora que tenemos el dataframe de TF-IDF, nos queda calcular la similitud coseno entre todos los vectores. Notar que la matriz resultante será una matriz simétrica. Implemente la función *cosine_similarity(v1, v2)* que recibe dos vectores (v1 y v2) y calcula la similitud coseno entre ambos vectores. Concluya cuáles son los dos documentos más similares.

In [98]:
def cosine_similarity(v1, v2):
  ### Aquí inicia tu código ###
  num = np.dot(v1, v2) #Producto punto entre los vectores
  den = np.linalg.norm(v1)*np.linalg.norm(v2) #Producto de la norma de los vectores
  return num/den
  ### Aquí termina tu código ### 

similarity_matrix = np.zeros((6,6))
for i, v1 in enumerate(tf_idf.index.values):
  for j, v2 in enumerate(tf_idf.index.values):
      similarity = cosine_similarity(tf_idf.loc[v1].values, tf_idf.loc[v2].values)
      similarity_matrix[i][j] = similarity
print(similarity_matrix)

[[1.         0.12032418 0.3223436  0.77967014 0.         0.1632828 ]
 [0.12032418 1.         0.08686457 0.         0.4952126  0.        ]
 [0.3223436  0.08686457 1.         0.1632828  0.         0.11787732]
 [0.77967014 0.         0.1632828  1.         0.12032418 0.3223436 ]
 [0.         0.4952126  0.         0.12032418 1.         0.08686457]
 [0.1632828  0.         0.11787732 0.3223436  0.08686457 1.        ]]


    Se puede notar que los documentos más similares en base a la similitud coseno son:
    - d0 = 'El perro se come la comida y después se duerme'
    - d3 = 'El gato se come la comida y después se duerme'

    Cuya similitud coseno es de 0.77967014. Lo cual tiene bastante sentido, ya que la única diferencia de las oraciones es el sujeto.


![gato](https://www.elagoradiario.com/wp-content/uploads/2020/05/Gato-mascarilla.jpg)

**Cualquier recomendación que nos quisieran dar para una futura tarea es bienvenid@!**